logpolicy/logpolicy.go

```diff
@@ -89,13 +89,9 @@ func (l logWriter) Write(buf []byte) (int, error) {
 }
 
 // New returns a new log policy (a logger and its instance ID) for a
-// given collection name. The provided filePrefix is used as a
-// filename prefix for both for the logger's state file, as well as
-// temporary log entries themselves.
-//
-// TODO: the state and the logs locations should perhaps be separated.
-func New(collection, filePrefix string) *Policy {
-	stateFile := filePrefix + ".log.conf"
+// given collection name. The logger persists its state to stateFile,
+// and buffers logs prior to upload in logsDir.
+func New(collection, stateFile, logsDir string) *Policy {
```

Comment:
Three positional args of same type isn't ideal at call sites.

Reply:
What's the alternative? Agreed that it's not ideal, although in practice it's invoked exactly once per binary in main(), so IMO not a massive deal.

Reply:
It seems like we could get this down to either 0 or 1 arguments: just the collection, if that couldn't even just be automatic based on the binary name. But probably just the collection name is easiest and most explicit.

The stateFile and logsDir seems like they could have automatic locations based on the user. For root, /var/log/*, else $HOME/var/tailscale/logs or something.

Reply:
The names I picked are not quite accurate, in the sense that `logsDir` actually stores temporary buffers. Their purpose is to buffer data prior to sending to log.tailscale.io, and to record crash goroutine dumps, so that the next process restart can pick it up and send it.

So in that sense it's more temp data than it is logs. We should still try to preserve it somewhat, but it's not extremely precious for long.

On that basis, system user location would be /var/cache/tailscale, state file would be /var/lib/tailscale (alongside tailscaled internal state). For users, ~/.cache/tailscale and ~/.config/tailscale.

Problem: we don't really want to autodetect "system user-ness". At some point I'm going to make tailscaled non-root, at which point the obvious "root is special" rule gets messed up. That's why I wanted some kind of way for the caller to decide where it's storing stuff.

Perhaps that means two constructors? `NewUser`, `NewSystem`? That internally call this `new` with appropriate paths?

Reply:
Note this logic also has to do the right thing on Windows and Mac. The os.* helpers you pointed out help with that, but we need to think that through also.

Reply:
`NewUser` and `NewSystem` is also not great for related reasons. That's you as a program author making assumptions about how/where it'll be run.

I'd just pick N OS-specific priority-ordered directories (where N=1 or 2) where you'd like to write state (keys, temp buffers) and then pick the first writable location.

Reply:
Scanning through dirs is still error-prone in a different way, in that multiple things may end up squashing each others's files :/

How about `New(collection, progName string)`, so that we can inject `progName` into filenames for differentiation?